subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/EdgeState.java

```diff
@@ -175,13 +176,29 @@ private void calculateTargetConfigurations() {
      */
     private void discoverAndApplyCapabilities(final ConfigurationMetadata targetConfiguration) {
         ImmutableList<? extends CapabilityDescriptor> capabilities = targetConfiguration.getCapabilities();
+        applyCapabilities(capabilities);
+    }
+
+    private void applyCapabilities(List<? extends CapabilityDescriptor> capabilities) {
         CapabilitiesHandlerInternal capabilitiesHandler = resolveState.getCapabilitiesHandler();
         for (final CapabilityDescriptor capability : capabilities) {
             final String prefer = capability.getPrefer();
             addCapabilityToResolutionState(capabilitiesHandler, capability, prefer);
         }
     }
 
+    private void applyCapabilitiesFromModule() {
+        // It is possible for a component which is _not_ selected to provide capabilities. They must be taken into
```

Comment:
May be not related, but does this means that if `org:foo:1.0` declares a capability but then gets upgraded to `org:foo:1.1` which no longer does, the capability will remain active?